### ISSUE TITLE
Deprecate Using the yegor256/rultor Docker Image for Other Projects

### ISSUE THREAD

**Issue by original-brownbear** [yegor256/rultor#1018]
Currently we have many of our own and I suspect also other projects running on the [default Docker image](https://hub.docker.com/r/yegor256/rultor/) for Rultor. 

**This is a significant risk to the stability of their builds.**

# Problem

Now what these projects have is this for their install section in the config file (or similar) this is from [Netbout](https://github.com/yegor256/netbout/blob/master/.rultor.yml):
```
  sudo gem install --no-ri --no-rdoc pdd
  sudo gem install --no-ri --no-rdoc est
```

For Rultor (and as a result those other projects) this means that Rultor has to:
1. Provide the sudo package/command and not stop doing so whenever a change to the Dockerfile is made, even if Rultor itself does not need the sudo command.
2. Not install any gems to the global scope that interfere with pdd or est
 * ... but also not install those two gems itself in a different way, say via RVM + separate Gemsets to prevent future incidents like https://github.com/yegor256/rultor/issues/1001, also causing harm to other projects.
* ... this is not a theoretical issue, this [error](https://github.com/yegor256/rultor/issues/994#issuecomment-179522451) was in many projects a while back, the fix made to it does not and cannot guarantee this not resurfacing again in a different form

While so far this only broke Gem related things, using the default image in other projects introduces risks elsewhere too, say in using the Rultor supplied Maven. For example we upgraded it to 3.3.9 from 3.2.1 in 
https://github.com/yegor256/rultor/pull/1005.
In this case I know of no backward compatibility problems (except for breaking Java 6 projects using Maven), but **what if the next time Rultor needs an upgrade to its Maven version and it is not compatible with what e.g. Netbout needs?**

# Fix Has to Provide:

1. A clear notice at the **very end** of the build output when building anything but Rultor itself while using the default `yegor256/rultor` image ( not providing your own ):
 * Please make it multiline :)
 * Link to http://doc.rultor.com/reference.html in that notice

2. A short explanation: 
 * clearly warning about the dangers of using Rultor without supplying your own Docker image/Dockerfile and hence falling back to `yegor256/rultor`
 * stating that running Rultor against the default image is deprecated.

**Comment by original-brownbear** [comment]
@alex-palevsky this is urgent.

**Comment by original-brownbear** [comment]
@alex-palevsky this is valid.

**Comment by original-brownbear** [comment]
@alex-palevsky this is a bug.

**Comment by original-brownbear** [comment]
@alex-palevsky this is a bug.

**Comment by alex-palevsky** [comment]
> @alex-palevsky this is a bug.

@original-brownbear tag "bug" added

**Comment by alex-palevsky** [comment]
@original-brownbear I set milestone here to `2.0`, let me know if it is wrong

**Comment by alex-palevsky** [comment]
@original-brownbear many thanks for the report, I topped your account for 30 mins, transaction AP-3E950352BE375454W

**Comment by original-brownbear** [comment]
@alex-palevsky this is urgent.

**Comment by alex-palevsky** [comment]
> @alex-palevsky this is urgent.

@original-brownbear right, I added "urgent" label

**Comment by alex-palevsky** [comment]
@gumbelmj this ticket is yours now, please proceed, and keep in mind [this](http://www.xdsd.org/2014/04/17/how-xdsd-is-different.html). Any technical questions you should ask right here; Task's budget is **30 mins** (see [this](http://www.xdsd.org/2014/04/17/how-xdsd-is-different.html) for explanation)

**Comment by gumbelmj** [comment]
@alex-palevsky  please assign to someone else

**Comment by alex-palevsky** [comment]
> @alex-palevsky  please assign to someone else

@gumbelmj I deducted 30 points from your rating

**Comment by alex-palevsky** [comment]
> @alex-palevsky  please assign to someone else

@gumbelmj no problem, I will try to find somebody else

**Comment by alex-palevsky** [comment]
@sebing help us, the task is yours

**Comment by sebing** [comment]
@original-brownbear Can you give some clarity in the work to be done?
- What to make multiline?
- Should I modify com.rultor.agents.docker.DockerExec to throw an exception if it is the default image(i have to figure it out what the mean myself)?

**Comment by original-brownbear** [comment]
@sebing sure, let my try to give a proper answer here :)

**What to make multiline?**
Just the notice at the bottom of a Rultor build, like say we look at this example log:
```
+ mv /home/r/repo .
++ whoami
+ chown -R root repo
+ '[' -n '' ']'
++ whoami
+ sudo chown -R rultor repo
+ cd repo
+ git push origin master
To git@github.com:someuser/somerepo.git
   af0fe22..e506072  master -> master
container 7d32415742c0fb09f4a5d87bd943c6725243c2c3256ed2139c56e3c21c1a9160 is dead
Fri Mar  4 12:18:13 UTC 2016
```
and assume the user had either no image or the `yegor256/rultor` image specified for his build in the `.rultor.yml`. Then the log should get a multi-line note appended, telling him that this is a "deprecated thing" to do ... along the lines of:
```
+ mv /home/r/repo .
++ whoami
+ chown -R root repo
+ '[' -n '' ']'
++ whoami
+ sudo chown -R rultor repo
+ cd repo
+ git push origin master
To git@github.com:someuser/somerepo.git
   af0fe22..e506072  master -> master
container 7d32415742c0fb09f4a5d87bd943c6725243c2c3256ed2139c56e3c21c1a9160 is dead
Fri Mar  4 12:18:13 UTC 2016
#### Deprecation Notice ####
You are using the Rultor default Docker image in your build.
.... some short version of above explanation as to why this is bad ....
########################
```

**Should I modify com.rultor.agents.docker.DockerExec to throw an exception if it is the default image(i have to figure it out what the mean myself)?**

No that class is not where you can do this, it is in fact only used to handle the cleaning of old images (and currently not really doing anything ... there's a ticket for that though.)
I suggested starting at `com.rultor.agents.daemons.StartsDaemon#run`, this function is what actually triggers the Docker run.
Also take a look at `com/rultor/agents/req/_head.sh` which is what eventually gets executed by the above function.
Lastly, this is not something that needs an exception. There is no error here that needs to be handled, it's just a user input issue that needs a simply but clear answer for the user :)

Let me know if this is enough to get you started here :)

**Comment by sebing** [comment]
@original-brownbear 
Thank you for the help. I think, I have a better understanding now.

- For the Deprecation Notice
Should I hardcode the details in `com/rulto/agents/daemons/end.sh`? And assuming `{image}` variable is set?

-Deprecating the default image.
`com/rultor/agents/req/_head.sh` If the `{image}` variable is `yegor256/rultor` i will echo a waring message?

**Comment by original-brownbear** [comment]
@sebing I think it would be best to put the logic for adding the notice into the Java code and just using and actually printing it from there. If we put it in bash it's a lot trickier to add a proper test for this.

**Comment by sebing** [comment]
@original-brownbear I don't understand. The logger output is not in the output of rultor log

**Comment by original-brownbear** [comment]
@sebing true, but sorry if this was unclear I did not refer to the logger :)
I think it's best of you give this a look too https://github.com/yegor256/rultor/blob/master/src/main/java/com/rultor/agents/daemons/Tail.java, this is the code that produces the output that Rultor always links to :)

**Comment by sebing** [comment]
@original-brownbear hmm... It is pretty confusing.
- In Tail.java the read method is returning an InputStream, is this stream is what we get in the output?(how is it possible for an inputstream in the output) may be I am not understanding the design at all.

- I see, `EndsDaemon` is responsible for the tail part of the message. It execute the `end.sh` so I felt the logic should somewhere here :)

**Comment by original-brownbear** [comment]
@sebing well you got it exactly right :) 

* Tail gives us the stream that is then html formatted and all that up the stack.
*  `com.rultor.agents.daemons.EndsDaemon#end` eventually produces the output, we're looking to hook into => go right ahead and handle it there :)

**Comment by sebing** [comment]
@original-brownbear Thanks again for the direction. 

Still I have not understood how or who holds the output in Java :(

**Comment by original-brownbear** [comment]
@sebing first off :) read this: http://www.yegor256.com/2015/01/15/how-to-cut-corners.html ,
as you can see this discussion has turned way too long and is not in line with the way we work.
In fact it hurts you the most, asking me for advice on the details of the implementation is not something that is rewarded, providing a clear and productive request for documentation on the other hand is clearly encouraged as it helps the project overall.

That being said, really `com.rultor.agents.daemons.EndsDaemon#end` is what you should care about.
It returns the `org.xembly.Directives` holding the output eventually.
So given the above, last and in fact a little more general tip here :)
I think a productive and more straight forwards approach to this would be to work from `com.rultor.agents.daemons.EndsDaemonITCase` (and potentially other tests) and understanding the code from this angle. I think that test clearly shows the path from what you appear to already understand to output string :)  If not, keep above link in mind ;)

**Comment by sebing** [comment]
@alex-palevsky The rultor project needs some documentation for the new guys jump into the project. At least a high level flow diagram.

**Comment by sebing** [comment]
@alex-palevsky I am trying to find a way to get the image in the `com.rultor.agents.daemons.EndsDaemon` and I cannot find any documentation for it. And I cannot understand the flow too. So can we create a task for creating documentation for this project?

**Comment by original-brownbear** [comment]
@sebing making tickets is not @alex-palevsky 's job, that's on me and you :)
I think #754 is what you're looking for here in general, obviously this one is a very general task and hence postponed for the time being.

I suggest creating a more specific issue asking for documentation on a certain area of Rultor, this is something that can be provided in a reasonable timeframe and everyone wins :)
In this case and in general, feel free to simply open issues of this kind, no need to ask for permission from anyone! 
I will review them shortly and we'll go from there :)

**Comment by sebing** [comment]
@original-brownbear Thanks for the information, I will keep them in mind.

**Comment by sebing** [comment]
@alex-palevsky  I have made a PR(https://github.com/yegor256/rultor/pull/1045)

**Comment by sebing** [comment]
@alex-palevsky I believe my solution to this issue is not up to the acceptable quality. 
- Should I wait for the documentation(https://github.com/yegor256/rultor/issues/754) to understand the project as per(http://www.yegor256.com/2015/01/15/how-to-cut-corners.html) 
- Or accept the temporal wrong solution I have in the PR(https://github.com/yegor256/rultor/pull/1045)
- Or assign the work to someone else :)

**Comment by sebing** [comment]
@alex-palevsky As I am taking too much time to understand the code without any documentation. May be you can assign the task to someone else?

**Comment by alex-palevsky** [comment]
> @alex-palevsky As I am taking too much time to understand the code without any documentation. May be you can assign the task to someone else?

@sebing -30 points to your rating :(

**Comment by alex-palevsky** [comment]
> @alex-palevsky As I am taking too much time to understand the code without any documentation. May be you can assign the task to someone else?

@sebing right, I'll try to find someone else for this task

**Comment by sebing** [comment]
@alex-palevsky The task is complete. Waiting for @original-brownbear to merge

**Comment by sebing** [comment]
@alex-palevsky as the the PR(https://github.com/yegor256/rultor/pull/1045) is merged. Can you close the issue?

**Comment by original-brownbear** [comment]
closing :) thanks!

**Comment by sebing** [comment]
@alex-palevsky So this will be considered as my work? ;)

**Comment by alex-palevsky** [comment]
@original-brownbear 2 puzzles were created here: `1018-15a5404e`, `1018-b83a728c`/#1052

**Comment by sebing** [comment]
@alex-palevsky As I finished the task, Did I get the point you have minus previously? Or that is the punishment for telling you that the code is difficult to understand ;)..

**Comment by sebing** [comment]
@alex-palevsky As this issue is closed and deployed. When will the rating will be updated? or I will end up with -30 :(

**Comment by alex-palevsky** [comment]
> @alex-palevsky As this issue is closed and deployed. When will the rating will be updated? or I will end up with -30 :(

@sebing no, this task is not yours that's why no points and no payments to you (if you disagree, talk to project architect)

**Comment by sebing** [comment]
@original-brownbear As I have finished this task in time. Do you think it is fair
- not get paid 
- and get a -30 points?

**Comment by original-brownbear** [comment]
@sebing I'm sorry but fair is not the category here in the first place. @alex-palevsky 's actions are in full compliance with the our policy. You asked to be reassigned, he reassigned you.
The fact that there was some delay here is pretty much irrelevant. 

**Still, regarding fair:**
I took a little time looking into this and found yegor256/takes#600 and associated pull request yegor256/takes#632.
As you can see this is not the first time the exact situation happened for you. In that ticket you also put in a PR after a ticket was reassigned already effectively ( you put it in  after you 48h deadline! ). 
**But** still the reviewer got paid there, even though his work did not at all contribute to the project,
given that the PR then needed to be closed since the ticket was reassigned.
=> Please understand this, the only way the way we work here is feasible is, if rules are strictly enforced. If spots like this one are not enforced, then the Architects and PMs in projects need to spend a lot of additional time working through situations like this one.
Since no one is paid for handling situations like the one at hand in a personal manner, we simply cannot have them. Me having answered this in detail was purely on my own time, so was you working on an issue you already resigned from.
I'm sorry, but this situation is in no way open to interpretation I think, please try to follow the rules more closely in the future.

**Comment by sebing** [comment]
@original-brownbear Thanks for the detailed reply it is great to know the detail always from you :).

- I have not understood the relation between (https://github.com/yegor256/rultor/issues/1018) and PR(https://github.com/yegor256/takes/pull/632) they are not connected. So I don't understand the point of bringing one of my previous mistakes here.
- The PR which is connected the issue(which I thought fair) is (https://github.com/yegor256/rultor/pull/1045). As this PR is accepted what is the problem for paying me?( I am not questioning your system just trying to understand it better so i won't do the same mistake again).

- According to my understanding what I did wrong is following this(http://www.yegor256.com/2015/01/15/how-to-cut-corners.html#say-no). The moment I realized, I may take time for an urgent task I informed @alex-palevsky It may take time for me to understand the code. If that is wrong I will never do that :)

-As you bring the issue (https://github.com/yegor256/takes/issues/600). I agree 100% it was my mistake. I can clarify my misunderstanding. @alice messaged me I have to do only 2 tasks {or that is what I have understood :)} So I kept the tasks in the pipeline and forget about it. As there is no direction, for the new joined team member. It is a bit difficult to know how we work here..(I don't think, my clarification is forgivable as it is chasing me all the task i do)

**Comment by sebing** [comment]
@original-brownbear @alex-palevsky Lesson learned here. Never say no. :)

**Comment by original-brownbear** [comment]
@sebing the lesson simply is, follow the rules without any exceptions ever.
You had two options here:
1. Asking for more time.
2. Saying no.
You said no (https://github.com/yegor256/rultor/issues/1018#issuecomment-194227106)
What you're doing when asking to give you a break here, is simply asking others to spent unpaid time for you, in order to sort these things out. If this would be done here, it would not be "fair" to not do this for everyone else and that would render rules pointless.

The only reason I brought the other mistake you made in Takes up, was to illustrate that these things cut both ways, **it was not in any way relevant even that you made the mistake there** (In fact I simply know about that one because I was waiting for the CR in there and was "stalking" him at the time ...).

... in the sense of http://www.yegor256.com/2014/04/20/how-hourly-rate-is-calculated.html#lengthy-discussions => just don't do it again, this issue is closed sorry.

**Comment by sebing** [comment]
@original-brownbear Okay i will keep quiet :) I won't be over-smart and assess the urgency of the issue anymore ;) 
So you fixed the issue without paying me, I understand it was my mistake to underestimate my ability that i can fix the issue. 

Thanks for the explanations.

**Comment by sebing** [comment]
@alex-palevsky @original-brownbear Just for the record, If anyone looking at the the treads. To know that you guys are co-coordinately working in "fairly"

- @alex-palevsky remove me from the issue after my request to remove for two days on `Mar 11, 2016, 1:52 PM GMT`
- The merge request is made in the pr at `Mar 11, 2016, 12:05 PM GMT`

So the manager knows theoretically the work is done and if I am removed from the task, I won't be needed to be paid.

**Comment by alex-palevsky** [comment]
> @alex-palevsky @original-brownbear Just for the record, If anyone looking at the the treads. To know that you guys are co-coordinately working in "fairly"
> 
> - @alex-palevsky remove me from the issue after my request to remove for two days on `Mar 11, 2016, 1:52 PM GMT`
> - The merge request is made in the pr at `Mar 11, 2016, 12:05 PM GMT`
> 
> So the manager knows theoretically the work is done and if I am removed from the task, I won't be needed to be paid.  

@sebing again, sorry about this

**Comment by alex-palevsky** [comment]
@original-brownbear the last puzzle `1053-dc86a5cc`/#1069 originated from here solved

**Comment by 0pdd** [comment]
@original-brownbear all 3 puzzles are solved here: [#1053](https://github.com/yegor256/rultor/issues/1053), [#1161](https://github.com/yegor256/rultor/issues/1161), [#unknown]().
